File: apps/embedding-service/schemas.py
from pydantic import BaseModel, Field, field_validator, model_validator

from settings import config


def validate_text_limits(values: list[str], field_name: str) -> None:
    # Character limits are request safety guards. Semantic chunking is token-based.
    total_chars = 0
    for index, value in enumerate(values):
        char_count = len(value)
        if char_count > config.embedding_max_text_chars:
            raise ValueError(
                f"{field_name}[{index}] exceeds maximum length of "
                f"{config.embedding_max_text_chars} characters"
            )
        total_chars += char_count
    if total_chars > config.embedding_max_request_chars:
        raise ValueError(
            f"{field_name} exceeds maximum total length of "
            f"{config.embedding_max_request_chars} characters"
        )


class EmbedRequest(BaseModel):
    texts: list[str] = Field(min_length=1)

    @field_validator("texts")
    @classmethod
    def validate_texts(cls, texts: list[str]) -> list[str]:
        if len(texts) > config.batch_limit:
            raise ValueError(f"too many texts; maximum batch size is {config.batch_limit}")
        for index, text in enumerate(texts):
            if not text or not text.strip():
                raise ValueError(f"texts[{index}] must not be empty")
        validate_text_limits(texts, "texts")
        return texts


class EmbeddedChunk(BaseModel):
    inputIndex: int
    chunkIndex: int
    chunkCount: int
    tokenCount: int
    text: str
    vector: list[float]


class EmbedResponse(BaseModel):
    model: str
    dimensions: int
    measuredTokens: int | None = None
    vectors: list[list[float]]
    chunks: list[EmbeddedChunk]


class RerankRequest(BaseModel):
    query: str = Field(min_length=1)
    documents: list[str] = Field(min_length=1)

    @field_validator("query")
    @classmethod
    def validate_query(cls, query: str) -> str:
        if not query.strip():
            raise ValueError("query must not be empty")
        validate_text_limits([query], "query")
        return query

    @field_validator("documents")
    @classmethod
    def validate_documents(cls, documents: list[str]) -> list[str]:
        if len(documents) > config.reranker_batch_limit:
            raise ValueError(
                f"too many documents; maximum batch size is {config.reranker_batch_limit}"
            )
        for index, document in enumerate(documents):
            if not document or not document.strip():
                raise ValueError(f"documents[{index}] must not be empty")
        validate_text_limits(documents, "documents")
        return documents

    @model_validator(mode="after")
    def validate_total_chars(self) -> "RerankRequest":
        validate_text_limits([self.query, *self.documents], "request")
        return self


class RerankResponse(BaseModel):
    model: str
    scores: list[float]

File: apps/embedding-service/src/schemas.ts
import type { EmbeddingServiceEnv } from "./env-config.js";
import { HttpError } from "./auth.js";

export interface EmbeddedChunk {
  inputIndex: number;
  chunkIndex: number;
  chunkCount: number;
  tokenCount: number;
  text: string;
  vector: number[];
}

export interface EmbedResponse {
  model: string;
  dimensions: number;
  measuredTokens: number | null;
  vectors: number[][];
  chunks: EmbeddedChunk[];
}

export interface RerankResponse {
  model: string;
  artifact: string;
  artifactRevision: string;
  artifactHash: string;
  latencyMs: number;
  inputTokens: number | null;
  costUsd: 0;
  scores: number[];
}

export const validateTextLimits = (
  config: Pick<
    EmbeddingServiceEnv,
    "embeddingMaxTextChars" | "embeddingMaxRequestChars"
  >,
  values: string[],
  fieldName: string
): void => {
  let totalChars = 0;
  values.forEach((value, index) => {
    const charCount = value.length;
    if (charCount > config.embeddingMaxTextChars) {
      throw new HttpError(
        422,
        `${fieldName}[${index}] exceeds maximum length of ${config.embeddingMaxTextChars} characters`
      );
    }
    totalChars += charCount;
  });
  if (totalChars > config.embeddingMaxRequestChars) {
    throw new HttpError(
      422,
      `${fieldName} exceeds maximum total length of ${config.embeddingMaxRequestChars} characters`
    );
  }
};

const stringArrayField = (
  payload: Record<string, unknown>,
  fieldName: string
): string[] => {
  const value = payload[fieldName];
  if (
    !Array.isArray(value) ||
    !value.every((item) => typeof item === "string")
  ) {
    throw new HttpError(422, `${fieldName} must be a non-empty string array`);
  }
  if (value.length === 0) {
    throw new HttpError(422, `${fieldName} must contain at least one item`);
  }
  return value;
};

export const validateEmbedRequest = (
  config: Pick<
    EmbeddingServiceEnv,
    "batchLimit" | "embeddingMaxTextChars" | "embeddingMaxRequestChars"
  >,
  payload: unknown
): { texts: string[] } => {
  if (
    typeof payload !== "object" ||
    payload === null ||
    Array.isArray(payload)
  ) {
    throw new HttpError(422, "request body must be a JSON object");
  }
  const texts = stringArrayField(payload as Record<string, unknown>, "texts");
  if (texts.length > config.batchLimit) {
    throw new HttpError(
      422,
      `too many texts; maximum batch size is ${config.batchLimit}`
    );
  }
  texts.forEach((text, index) => {
    if (!text || !text.trim()) {
      throw new HttpError(422, `texts[${index}] must not be empty`);
    }
  });
  validateTextLimits(config, texts, "texts");
  return { texts };
};

export const validateRerankRequest = (
  config: Pick<
    EmbeddingServiceEnv,
    "rerankerBatchLimit" | "embeddingMaxTextChars" | "embeddingMaxRequestChars"
  >,
  payload: unknown
): { query: string; documents: string[] } => {
  if (
    typeof payload !== "object" ||
    payload === null ||
    Array.isArray(payload)
  ) {
    throw new HttpError(422, "request body must be a JSON object");
  }
  const record = payload as Record<string, unknown>;
  if (typeof record.query !== "string") {
    throw new HttpError(422, "query must be a string");
  }
  if (!record.query.trim()) {
    throw new HttpError(422, "query must not be empty");
  }
  validateTextLimits(config, [record.query], "query");

  const documents = stringArrayField(record, "documents");
  if (documents.length > config.rerankerBatchLimit) {
    throw new HttpError(
      422,
      `too many documents; maximum batch size is ${config.rerankerBatchLimit}`
    );
  }
  documents.forEach((document, index) => {
    if (!document || !document.trim()) {
      throw new HttpError(422, `documents[${index}] must not be empty`);
    }
  });
  validateTextLimits(config, documents, "documents");
  validateTextLimits(config, [record.query, ...documents], "request");
  return { query: record.query, documents };
};
